aten/src/ATen/Context.h

```diff
@@ -488,23 +489,36 @@ class TORCH_API Context {
   bool enable_sparse_tensor_invariant_checks = false;
   bool allow_fp16_reduction_cpu = false;
 
-  std::map<std::string, std::map<std::string, std::string>> fp32_precision = {
-      {"generic", {{"all", "none"}}},
-      {"mkldnn",
-       {{"matmul", "none"},
-        {"conv", "none"},
-        {"rnn", "none"},
-        {"all", "none"}}},
-      {"cuda",
-       {{"matmul",
-         float32_matmul_precision == at::Float32MatmulPrecision::HIGHEST
-             ? "none"
-             : "tf32"},
-        {"conv", "tf32"},
-        {"rnn", "tf32"},
-        {"all", "none"}}},
+  // manually designed hash function
+  struct BackendOpHash {
+    size_t operator()(const std::pair<std::string, std::string>& p) const {
+      if (C10_UNLIKELY(p.first.empty() || p.second.empty())) {
+        return 0ull;
+      }
+      return 2 * static_cast<size_t>(p.first[0]) +
```

Comment:
this is super unsafe, what if at some point we have keys that start with the same letter? Can we switch to enums to avoid this? As was discussed in the previous issues, they, for some weird reason, want string on the python side, fine, but on c++ we should just have enums.

Reply:
Yes we should change all the callsites in c++.